algorithms/src/Kokkos_Random.hpp

```diff
@@ -903,24 +903,45 @@ class Random_XorShift64_Pool {
 #else
   Random_XorShift64_Pool()   = default;
 #endif
+
   Random_XorShift64_Pool(uint64_t seed) {
-    num_states_ = 0;
+    init(execution_space(), seed, execution_space().concurrency());
+    execution_space().fence();
+  }
+
+  Random_XorShift64_Pool(uint64_t seed, uint64_t num_states) {
+    init(execution_space(), seed, num_states);
+    execution_space().fence();
+  }
 
-    init(seed, execution_space().concurrency());
+  Random_XorShift64_Pool(const execution_space& exec, uint64_t seed) {
+    init(exec, seed, exec.concurrency());
   }
 
-  void init(uint64_t seed, int num_states) {
+  Random_XorShift64_Pool(const execution_space& exec, uint64_t seed,
+                         uint64_t num_states) {
+    init(exec, seed, num_states);
+  }
+
+  void init(uint64_t seed, uint64_t num_states) {
+    init(execution_space(), seed, num_states);
```

Comment:
Now we have changed the behavior of the member function to async.

Reply:
yeah, we should pull the fence from the ctors to here